### PR TITLE
041 login system UI test

### PR DESCRIPTION
**Pull request by henrykainos**

### CODE REVIEW

File: src/test/unit/ui/HomeTestPage.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { By, WebDriver } from 'selenium-webdriver';
+
+export class HomeTestPage {
+    private driver: WebDriver;
+    
+    // Locators
```
Comment (boydstepkainos) [inline src/test/unit/ui/HomeTestPage.ts]: Can delete the comments

File: src/test/unit/ui/LoginTest.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import { Builder, Capabilities, WebDriver } from 'selenium-webdriver';
+import { expect } from 'chai';
+import { LoginTestPage } from "./LoginTestPage";
+import { LogoutTestPage } from "./LogoutTestPage";
+import { HomeTestPage } from "./HomeTestPage";
+import { NavbarTestPage } from "./NavbarTestPage";
+import * as chrome from 'selenium-webdriver/chrome';
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: import only what you're using

File: src/test/unit/ui/LoginTestPage.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// src/pages/LoginTestPage.ts
+import { By, WebDriver } from 'selenium-webdriver';
+
+export class LoginTestPage {
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTestPage.ts]: I like the separation of these elements for reuse, however I'd create a new directory for them to keep them separate from actual tests

File: src/test/unit/ui/LoginTest.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+
+describe('Login Test', function () {
+    this.timeout(10000) //Setting max timeout, default 2s
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: Why max?

File: src/test/unit/ui/LoginTest.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    let driver: WebDriver;
+    let loginPage: LoginTestPage;
+    let logoutPage: LogoutTestPage;
+    let homePage: HomeTestPage;
+    let navBarPage: NavbarTestPage;
+    before(async function () {
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: newline to separate function from variables

File: src/test/unit/ui/LoginTest.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        loginPage = new LoginTestPage(driver);
+        logoutPage = new LogoutTestPage(driver);
+        homePage = new HomeTestPage(driver);
+        navBarPage = new NavbarTestPage(driver);
+
+    });
+
+    after(async function () {
+        try {
+            await driver.quit(); // Ensure this is awaited
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: can remove comment

File: src/test/unit/ui/LoginTest.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        } catch (error) {
+            console.error('Error quitting the driver:', error);
+        }
+    });
+
+    it('Should successfully login and logout with valid email and password', async function () {
+        const url: string = process.env.UI_TEST_URL || 'https://nczcbkjcc7.eu-west-1.awsapprunner.com/loginForm';
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: set this as variable in before

File: src/test/unit/ui/LoginTest.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        await loginPage.open(url);
+
+        await loginPage.enterEmail('adam@random.com');
+        await loginPage.enterPassword('pass123');
+        await loginPage.clickLogin();
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: is there nothing to assert on login?

File: src/test/unit/ui/LoginTest.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        await loginPage.enterEmail('adam123@random.com');
+        await loginPage.enterPassword('pass123');
+        await loginPage.clickLogin();
+        
+        const actualText = await loginPage.getErrorMessageText();
+        expect(actualText).to.equal('user does not exist.');
+        
+    });
+
+    it('Should fail login with valid email and invalid pasword', async function () {
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: typo in password

File: src/test/unit/ui/LoginTest.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        await loginPage.enterEmail('');
+        await loginPage.enterPassword('');
+        await loginPage.clickLogin();
+        
+        const actualText = await loginPage.getErrorMessageText();
+        expect(actualText).to.equal('user does not exist.');
+        
+    });
+
+    it('Should fail login with invalid email and pasword', async function () {
```
Comment (boydstepkainos) [inline src/test/unit/ui/LoginTest.ts]: same typo

File: src/test/unit/ui/HomeTestPage.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    private titleMessage: By = By.xpath("/html/body/div[1]/div/h2");
```
Comment (boydstepkainos) [inline src/test/unit/ui/HomeTestPage.ts]: this relies on the page never changing, use an id or something consistent

Comment (boydstepkainos) [reply src/test/unit/ui/HomeTestPage.ts]: do this for all selectors

File: src/test/unit/ui/LoginTest.ts (hunk 2 above)
Comment (henrykainos) [reply src/test/unit/ui/LoginTest.ts]: All these imports are being used

File: src/test/unit/ui/LoginTest.ts (hunk 4 above)
Comment (henrykainos) [reply src/test/unit/ui/LoginTest.ts]: The test by default timesout if it takes longer than 2 seconds, this code increases the amount of time until before a timeout is triggered

File: src/test/unit/ui/HomeTestPage.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    constructor(driver: WebDriver) {
+        super(driver);
+    }
+
+    async open(theURL:string): Promise<void> {
+        await this.go_to_url(theURL);
+    }
+
+    async getTitleText(): Promise<string> {
+        const element = '/html/body/div[1]/div/h2'; //Placeholder, waiting for hotfix for better solution
```
Comment (shaunganley) [inline src/test/unit/ui/HomeTestPage.ts]: get by ID rather than xpath

File: src/test/unit/ui/LoginTest.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    after(async function () {
+        try {
+            await loginPage.closeBrowser();
+        } catch (error) {
+            console.error('Error quitting the driver:', error);
+        }
+    });
+
+    it('Should successfully login and logout with valid email and password', async function () {
+        await loginPage.open("https://nczcbkjcc7.eu-west-1.awsapprunner.com/loginForm");
```
Comment (shaunganley) [inline src/test/unit/ui/LoginTest.ts]: pass the base URL in as an environment variable so these can be run locally too

File: src/test/unit/ui/LoginTest.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+        await loginPage.enterEmail('adam@random.com');
+        await loginPage.enterPassword('pass123');
```
Comment (shaunganley) [inline src/test/unit/ui/LoginTest.ts]: don't hardcode valid login creds in code. pass in as environment variables

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — DISMISSED** [review]

File: src/test/ui/NavbarTestPage.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        const element = 'Home';
+        await this.clickByLinkText(element);
+    }
+    async clickJobsButton(): Promise<void> {
+        const element = 'Jobs';
+        await this.clickByLinkText(element);
+    }
+    async clickAboutUsButton(): Promise<void> {
+        const element = 'About Us';
+        await this.clickByLinkText(element);
```
Comment (kajatraczykowska) [inline src/test/ui/NavbarTestPage.ts]: About us page doesn't exist and button has been removed from the navbar so no need to test it

File: src/test/ui/LoginTest.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+import { WebDriver } from 'selenium-webdriver';
+import { ChromeDriver } from './ChromeDriver';
+
+describe('Login Test', function () {
+    this.timeout(30000) //Setting max timeout so test doesnt timeout after 2s
+
+    let driver: WebDriver;
+    let loginPage: LoginTestPage;
+    let logoutPage: LogoutTestPage;
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (kajatraczykowska) [inline src/test/ui/LoginTest.ts:15]: is this comment necessary?

**Review by kajatraczykowska — CHANGES_REQUESTED** [review]

File: src/test/ui/LoginTest.ts (hunk 16 above)
Comment (henrykainos) [reply src/test/ui/LoginTest.ts:15]: For linter

**Review by jackpollinsK — DISMISSED** [review]

**Review by AdamB-Kainos — APPROVED** [review]

**Review by AdamB-Kainos — DISMISSED** [review]

**Review by kajatraczykowska — APPROVED** [review]

File: .github/workflows/build.yml [hunk 17]
```diff
@@ -19,13 +19,9 @@ jobs:
...
       uses: actions/setup-node@v2
       with:
-        node-version: 20.14.0
+        node-version: 20.14.0   
 
     - name: Install dependencies
       run: npm install
 
-    - name: Run unit tests
-      run: npm test
```
Comment (shaunganley) [inline .github/workflows/build.yml:28]: leave this in here

File: .github/workflows/uiTests.yml [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+    - name: Set up Node.js
+      uses: actions/setup-node@v2
+      with:
+        node-version: 20.14.0
+    
+    - name: Install dependencies
+      run: npm install
+  
+    - name: Install Google Chrome
```
Comment (shaunganley) [inline .github/workflows/uiTests.yml:23]: can't you just run the ui tests in headless mode?

Comment (henrykainos) [reply .github/workflows/uiTests.yml:23]: They are but had to install specific version of chrome compatible with chromedriver

**Review by shaunganley — APPROVED** [review]

**Pull request merged**